contracts/staking/contracts/src/fees/MixinExchangeFees.sol

```diff
@@ -250,4 +284,69 @@ contract MixinExchangeFees is
             finalContractBalance
         );
     }
+
+    /// @dev The cobb-douglas function used to compute fee-based rewards for staking pools in a given epoch.
+    /// Note that in this function there is no limitation on alpha; we tend to get better rounding
+    /// on the simplified versions below.
+    /// @param totalRewards collected over an epoch.
+    /// @param ownerFees Fees attributed to the owner of the staking pool.
+    /// @param totalFees collected across all active staking pools in the epoch.
+    /// @param ownerStake Stake attributed to the owner of the staking pool.
+    /// @param totalStake collected across all active staking pools in the epoch.
+    /// @param alphaNumerator Numerator of `alpha` in the cobb-dougles function.
+    /// @param alphaDenominator Denominator of `alpha` in the cobb-douglas function.
+    /// @return ownerRewards Rewards for the owner.
+    function _cobbDouglas(
+        uint256 totalRewards,
+        uint256 ownerFees,
+        uint256 totalFees,
+        uint256 ownerStake,
+        uint256 totalStake,
+        uint256 alphaNumerator,
+        uint256 alphaDenominator
+    )
+        internal
+        pure
+        returns (uint256 ownerRewards)
+    {
+        assert(alphaNumerator <= alphaDenominator);
+        int256 feeRatio = LibFixedMath._toFixed(ownerFees, totalFees);
+        int256 stakeRatio = LibFixedMath._toFixed(ownerStake, totalStake);
+        if (feeRatio == 0 || stakeRatio == 0) {
+            return ownerRewards = 0;
+        }
+
+        // The cobb-doublas function has the form:
+        // totalRewards * feeRatio ^ alpha * stakeRatio ^ (1-alpha)
+        // We instead use:
+        // totalRewards * stakeRatio * e^(alpha * (ln(feeRatio) - ln(stakeRatio)))
+
+        // Compute e^(alpha * (ln(feeRatio) - ln(stakeRatio)))
+        int256 logFeeRatio = LibFixedMath._ln(feeRatio);
```

Comment:
Do you have a good handle on how the error of `_ln(x)` varies with `x`? 
Given that we get fewer digits of precision in the  `ln(x), where x is close to 0` test, I assume it's more numerically stable for `x` closer to 1 –– if that's the case, we might be able to squeeze out a bit more precision using the following observation:
`ln(F_i / F) - ln(S_i / S))`, `ln(F_i / R_i) - ln(F / R)`, `-ln(R_i / F_i) - ln(F / R)`, `ln(R / F) + ln(F_i / R_i)`, `ln(R / F) - ln(R_i / F_i)` are all equivalent. 
For given values of `F_i, R_i, F, R`, only one of {`ln(F_i / R_i) - ln(F / R)`, `-ln(R_i / F_i) - ln(F / R)`, `ln(R / F) + ln(F_i / R_i)`, `ln(R / F) - ln(R_i / F_i)`} will work with `_ln`, but it might be more stable than `ln(F_i / F) - ln(S_i / S))`. 
So we can do some quick comparisons of `F_i, R_i, F, R` to get a heuristic of which version will be more precise. This is all a theoretical hunch though, would need testing.

Reply:
I guess we could also potentially use `ln((F_i * R) / (F * R_i))` or `-ln((F * R_i) / (F_i * R))`

Reply:
It's way too easy to bait me into a rabbit hole:

![ln_error](https://drive.google.com/uc?export=download&id=1xbl0hczLa4_AmmlKfuQlmQ7Fvg8PtdAS)

Reply:
It's funny because I originally had your suggestion of `ln(x = (F_i * R) / (F * R_i))` but scrapped it since `x` would be unbounded. Then I added in the conditional inverse but completely forgot that I could now switch back. I'll make the change and run before and after precision tests.

Reply:
😆 that error curve is jank

Reply:
You wanna see jank? Check out the error on the native float64 implementation we all know and love.:

![lnx_native_vs_solidity](https://drive.google.com/uc?export=download&id=1xJKSEGhMk3N5NUje3lKIkZmKgXcPKSzo)

Reply:
So I played around with adding more high order terms to `ln()` to squeeze out more precision. It was pretty negligible any way I sliced it. The huge error spike towards `1` is probably due to the taylor series not having enough terms, so I can look into generating more of those, but I'm not sure if it's worth the extra gas, since I think. in practice,  we won't see inputs near `1.0` if the market is well diversified. Also, 14 digits of precision in these cases is probably good enough, I think?

I also did end up settling on the two following cobb-douglas forms, which **saves >1k gas** and left precision unchanged. I also refactored it somewhat to hopefully reduce code gen.
```python
totalRewards * stakeRatio * e^(alpha * (ln(feeRatio / stakeRatio))) # if feeRatio <= stakeRatio
totalRewards * stakeRatio / e^(alpha * (ln(stakeRatio / feeRatio))) # if feeRatio > stakeRatio
```
This is because `ln(x)` has domain `0 < x <= 1` , `exp(x)` has domain `x < 0`, and 127-bit fixed-point overflows quite easily when multiplying numbers > ~2.